include/extractor/guidance/intersection_optimizer.hpp

```diff
@@ -0,0 +1,101 @@
+#ifndef OSRM_EXTRACTOR_GUIDANCE_INTERSECTION_OPTIMIZER_HPP_
+#define OSRM_EXTRACTOR_GUIDANCE_INTERSECTION_OPTIMIZER_HPP_
+
+#include "util/typedefs.hpp"
+
+#include "util/attributes.hpp"
+
+#include "extractor/guidance/intersection.hpp"
+#include "extractor/guidance/intersection_generator.hpp"
+#include "extractor/guidance/coordinate_extractor.hpp"
+#include "extractor/query_node.hpp"
+
+#include "extractor/suffix_table.hpp"
+#include "util/name_table.hpp"
+
+#include <vector>
+
+namespace osrm
+{
+namespace extractor
+{
+namespace guidance
+{
+
+/*
+ * An intersection is a central part in computing guidance decisions. However the model in OSM and
+ * the view we want to use in guidance are not necessarily the same thing. We have to account for
+ * some models that are chosen explicitly in OSM and that don't actually describe how a human would
+ * experience an intersection.
+ *
+ * For example, if a small pedestrian island is located at a traffic light right in the middle of a
+ * road, OSM tends to model the road as two separate ways. A human would consider these two ways a
+ * single road, though. In this optimizer, we try to account for these subtle differences between
+ * OSM data and human perception to improve our decision base for guidance later on.
+ */
+class IntersectionOptimizer
+{
+  public:
+    IntersectionOptimizer(const util::NodeBasedDynamicGraph &node_based_graph,
+                          const std::vector<extractor::QueryNode> &node_coordinates,
+                          const util::NameTable &name_table,
+                          const SuffixTable &street_name_suffix_table,
+                          const IntersectionGenerator &intersection_generator);
+
+    // The function takes an intersection an converts it to a `perceived` intersection which closer
+    // represents how a human might experience the intersection
+    OSRM_ATTR_WARN_UNUSED
+    Intersection Process(const NodeID node_at_intersection, Intersection intersection) const;
```

Comment:
What's your stance on operator() e.g. in the handlers and having a named single function here?